File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Predict.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.examples.vnni.bigdl

import com.intel.analytics.zoo.common.NNContext
import com.intel.analytics.zoo.feature.image.ImageSet
import com.intel.analytics.zoo.models.image.imageclassification.{ImageClassifier, LabelOutput}
import org.apache.log4j.{Level, Logger}
import org.opencv.imgcodecs.Imgcodecs
import scopt.OptionParser

case class PredictParams(folder: String = "./",
                         model: String = "",
                         topN: Int = 5)

object Predict {
  Logger.getLogger("com.intel.analytics.bigdl.transform.vision").setLevel(Level.ERROR)
  Logger.getLogger("com.intel.analytics.zoo.feature.image").setLevel(Level.ERROR)

  val logger: Logger = Logger.getLogger(getClass)

  def main(args: Array[String]): Unit = {
    System.setProperty("bigdl.engineType", "mkldnn")

    val parser = new OptionParser[PredictParams]("ResNet50 Int8 Inference Example") {
      opt[String]('f', "folder")
        .text("The local folder path that contains images for prediction")
        .action((x, c) => c.copy(folder = x))
        .required()
      opt[String]('m', "model")
        .text("The path to the downloaded int8 model snapshot")
        .action((x, c) => c.copy(model = x))
        .required()
      opt[Int]("topN")
        .text("The top N classes with highest probabilities as output")
        .action((x, c) => c.copy(topN = x))
    }
    parser.parse(args, PredictParams()).map(param => {
      val sc = NNContext.initNNContext(
        "Image classification inference example using int8 quantized model")
      val images = ImageSet.read(param.folder, imageCodec = Imgcodecs.CV_LOAD_IMAGE_COLOR)
      val model = ImageClassifier.loadModel[Float](param.model)
      logger.info(s"Start inference on images under ${param.folder}...")
      val output = model.predictImageSet(images)
      val labelOutput = LabelOutput(model.getConfig().labelMap, probAsOutput = false)
      val results = labelOutput(output).toLocal().array

      logger.info(s"Prediction results:")
      results.foreach(imageFeature => {
        logger.info(s"image: ${imageFeature.uri}, top ${param.topN}")
        val classes = imageFeature("classes").asInstanceOf[Array[String]]
        val probs = imageFeature("probs").asInstanceOf[Array[Float]]
        for (i <- 0 until param.topN) {
          logger.info(s"\t class: ${classes(i)}, credit: ${probs(i)}")
        }
      })
      logger.info(s"Prediction finished.")
      sc.stop()
    })
  }
}

File: pyzoo/zoo/examples/vnni/bigdl/Predict.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys
from zoo.models.image.imageclassification import ImageClassifier
from zoo.pipeline.api.keras.layers import Activation
from zoo.pipeline.api.keras.models import Sequential
from zoo.feature.image import ImageSet
from zoo.common.nncontext import init_nncontext
from optparse import OptionParser


def predict(model_path, image_path, top_n):
    sc = init_nncontext("Image classification inference example using int8 quantized model")
    images = ImageSet.read(image_path, sc, image_codec=1)
    model = ImageClassifier.load_model(model_path)
    output = model.predict_image_set(images)
    label_map = model.get_config().label_map()

    # list of images composing uri and results in tuple format
    predicts = output.get_predict().collect()

    sequential = Sequential()
    sequential.add(Activation("softmax", input_shape=predicts[0][1][0].shape))
    for pre in predicts:
        (uri, probs) = pre
        out = sequential.forward(probs[0])
        sortedProbs = [(prob, index) for index, prob in enumerate(out)]
        sortedProbs.sort()
        print("Image : %s, top %d prediction result" % (uri, top_n))
        for i in range(top_n):
            print("\t%s, %f" % (label_map[sortedProbs[999 - i][1]], sortedProbs[999 - i][0]))


if __name__ == "__main__":
    parser = OptionParser()
    parser.add_option("--model", type=str, dest="model_path",
                      help="The path to the downloaded int8 model snapshot")
    parser.add_option("--image", type=str, dest="image_path",
                      help="The local folder path that contains images for prediction")
    parser.add_option("--topN", type=int, dest="topN", default=5,
                      help="The top N classes with highest probabilities as output")

    (options, args) = parser.parse_args(sys.argv)
    predict(options.model_path, options.image_path, options.topN)
